Model.py

```diff
@@ -0,0 +1,429 @@
+
+        Returns:
+            x: Compressed and sliced batch: torch tensor of size (batch_size x new_n_sample_points x slice_size).
+        """
+        
+        # Apply projection
+        x = self.projection(x)
+        return x
+    
+
+    
+""" ********** Multi-head attention ********** """
+
+class MultiHeadAttention(nn.Module):
+    
+    def __init__(self, slice_size, num_heads, dropout):
+        
+        """   
+        Warning: num_heads must be a divider of slice_size !
+        Args:
+            slice_size (int): Size of slices,
+            num_heads (int): Number of heads in mulit-head block,
+            dropout (float): Dropout value for multi-head block,.
+        """
+        
+        super().__init__()
+        self.scaling = (slice_size / num_heads) ** (1 / 2)
+        self.num_heads = num_heads
+        self.keys = nn.Linear(slice_size, slice_size)
+        self.queries = nn.Linear(slice_size, slice_size)
+        self.values = nn.Linear(slice_size, slice_size)
+        self.att_drop = nn.Dropout(dropout)
+        self.projection = nn.Linear(slice_size, slice_size)
+
+    def forward(self, x: Tensor, mask: Tensor = None) -> Tensor:
+       
+        """
+        Apply mulit-head attention on compressed and sliced batch x of size (batch_size x new_n_sample_points x slice_size).
+
+        Args:
+            x (torch tensor): Batch of trial after compression and slicing (batch_size x new_n_sample_points x slice_size).
+
+        Returns:
+             out: Torch tensor of size (batch_size x new_n_sample_points x slice_size).
+        """
+        
+        # Compute Q, K, V
+        queries = rearrange(self.queries(x), 'b q (h o) -> b h q o', h = self.num_heads) # q = query length
+        keys = rearrange(self.keys(x), 'b k (h o) -> b h k o', h = self.num_heads) # k = key length
+        values = rearrange(self.values(x), 'b v (h o) -> b h v o', h = self.num_heads) # v = value length and q = k = v
+        
+        # Compute attention score
+        energy = torch.einsum('b h q o, b h k o -> b h q k', queries, keys)  # batch, num_heads, query_len, key_len
+        if mask is not None:
+            fill_value = torch.finfo(torch.float32).min 
+            energy.mask_fill(~mask, fill_value)
+
+        attention_score = F.softmax(energy / self.scaling, dim=-1) 
+        attention_score = self.att_drop(attention_score)
+        
+        # Multiply by value V 
+        out = torch.einsum('b h q k, b h v o -> b h v o ', attention_score, values) # q = v
+        
+        # Apply projection
+        out = rearrange(out, "b h v o -> b v (h o)")
+        out = self.projection(out)
+        return out
+    
+    
+    
+""" ********** Feed-forward block ********** """
+
+class GELU(nn.Module):
+    
+    """
+    GeLU activation function.
+    """
+    
+    def forward(self, input: Tensor) -> Tensor:
+        return input*0.5*(1.0+torch.erf(input/math.sqrt(2.0)))
```

Comment:
I am not familiar with GELU.

A quick look at the paper describing it:
https://arxiv.org/pdf/1606.08415v3.pdf
shows a different formula.

Please confirm yours gets same/approximate values.

Reply:
This is what is seemed to me: in the model, nn.GELU() is used (GELU function coded in Pytorch) so the GELU function implemented is not used. I erased it as it is useless